synse/http.go

```diff
@@ -0,0 +1,152 @@
+	client := resty.New()
+	err := setupHTTPClient(options, client)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to setup a http client")
+	}
+
+	return &httpClient{
+		options: options,
+		client:  client,
+	}, nil
+}
+
+// setupHTTPClient setups the client with configured options.
+func setupHTTPClient(opt *Options, c *resty.Client) error {
+	if opt == nil {
+		return errors.New("options can not be nil")
+	}
+
+	if opt.Server.Address == "" {
+		return errors.New("no address is specified")
+	}
+	c = c.SetHostURL(fmt.Sprintf("http://%s/synse/", opt.Server.Address))
+
+	if opt.Server.Timeout == 0 {
+		// FIXME - find a better way to use default options here?
+		opt.Server.Timeout = 2 * time.Second
+		c = c.SetTimeout(opt.Server.Timeout)
+	}
+
+	// Only use retry options if set, otherwise let the resty client goes with
+	// its defaults (O Count, 100 milliseconds WaitTime, 2 seconds MaxWaitTime).
+	if opt.Retry.Count != 0 {
+		c = c.SetRetryCount(opt.Retry.Count)
+	}
+
+	if opt.Retry.WaitTime != 0 {
+		c = c.SetRetryWaitTime(opt.Retry.WaitTime)
+	}
+
+	if opt.Retry.MaxWaitTime != 0 {
+		c = c.SetRetryMaxWaitTime(opt.Retry.MaxWaitTime)
+	}
+
+	return nil
+}
+
+// Status returns the status info.
+func (c *httpClient) Status() (*Status, error) {
+	out := new(Status)
+	synseError := new(Error)
+
+	err := c.getUnversioned(testURI, out, synseError)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to request `/status` endpoint")
+	}
+
+	return out, nil
+}
+
+// Version returns the version info.
+func (c *httpClient) Version() (*Version, error) {
+	out := new(Version)
+	synseError := new(Error)
+
+	err := c.getUnversioned(versionURI, out, synseError)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to request `/version` endpoint")
+	}
+
+	return out, nil
+}
+
+// Config returns the config info.
+func (c *httpClient) Config() (*Config, error) {
+	out := new(Config)
+	synseError := new(Error)
+
+	err := c.getVersioned(configURI, out, synseError)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to request `/config` endpoint")
+	}
+
+	return out, nil
+}
+
+// getUnversioned makes an unversioned request.
+func (c *httpClient) getUnversioned(path string, okResp interface{}, errResp *Error) error {
+	_, err := c.client.R().
+		SetResult(okResp).
+		SetError(errResp).
+		Get(path)
+	if err != nil {
+		return errors.Wrap(err, "failed to make an unversioned request to synse server")
+	}
+
+	if *errResp != (Error{}) {
+		return errors.Errorf("got an error response from synse server: %+v", errResp)
+	}
+
+	return nil
+}
+
+// getVersioned makes a versioned request.
+func (c *httpClient) getVersioned(path string, okResp interface{}, errResp *Error) error {
+	err := c.cacheAPIVersion()
+	if err != nil {
+		return errors.Wrap(err, "failed to cache api version")
+	}
+
+	versionedPath := fmt.Sprintf("%s/%s", c.apiVersion, path)
+	err = c.getUnversioned(versionedPath, okResp, errResp)
```

Comment:
Another thought: we could have different constructors for clients, e.g. (pseudocode)

```go
// Creates a new synse client, does not specify API version.
func NewSynseClient() (*Client, error)

// Creates a new synse client with API v3
func NewSynseV3Client() (*Client, error)
```

Then, when it comes to URL construction we have some options:
- if the client has a version specified (e.g. "v3", use it)
- if the client does not have a version specified, look it up using the `/version` endpoint.

This is ultimately a larger bit of work as it relates to backwards compatibility, so it doesn't need to be addressed now, but we should definitely open an issue for it.